File: platform/macos/Storage/NativeTrustStore.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import CoreFoundation
import CryptoKit
import Foundation
import TidyVNC

public enum NativeTrustKind: String, Sendable {
  case certificate = "x509-spki", hostKey = "rsa-aes"
  var domain: String { "io.github.jkeli.tidyvnc.trust." + rawValue + ".v1" }
  public func savedFingerprintMessage(_ fingerprint: String) -> String {
    switch self {
    case .certificate: String(localized:"trust.library.saved.spki", defaultValue:"Saved SPKI SHA-256: \(fingerprint)")
    case .hostKey: String(localized:"trust.library.saved.serverKey", defaultValue:"Saved Server-key SHA-256: \(fingerprint)")
    }
  }
}
// Domain-separated canonical endpoint identity. Original labels are retained only
// for management UI; stored scopes are rederived on load, never trusted as labels.
public struct NativeTrustScope: Hashable, Sendable, CustomStringConvertible, CustomDebugStringConvertible {
  public let id: String, endpoint: String, routeIdentity: String
  public let kind: NativeTrustKind
  public var description: String { "NativeTrustScope(<redacted>)" }
  public var debugDescription: String { description }
  public static func ==(lhs: Self, rhs: Self) -> Bool { lhs.id == rhs.id }
  public func hash(into hasher: inout Hasher) { hasher.combine(id) }
  public init(endpoint: String, routeIdentity: String = "", kind: NativeTrustKind = .certificate) throws {
    func bounded(_ value: String) throws -> [UInt8] {
      let result = Array(value.utf8.prefix(4097))
      guard result.count <= 4096, !result.contains(0) else { throw NativeStorageError.invalid }
      return result
    }
    let address = try bounded(endpoint), route = try bounded(routeIdentity)
    guard !address.isEmpty else { throw NativeStorageError.invalid }
    var raw: UInt64 = 0
    _ = try address.withUnsafeBufferPointer { address in
      try route.withUnsafeBufferPointer { route in
        try checked { tidyvnc_endpoint_create(.init(data: address.baseAddress,length: UInt64(address.count)),
          .init(data: route.baseAddress,length: UInt64(route.count)),1,&raw,$0) }
      }
    }
    let owner = NativeHandle(adopting: raw)
    id = try withExtendedLifetime(owner) {
      var value = abi(tidyvnc_endpoint_info.self)
      try checked { tidyvnc_endpoint_get(owner.raw,&value,$0) }
      var hash = SHA256()
      func number(_ value: UInt32) -> Data { var big = value.bigEndian; return withUnsafeBytes(of: &big) { Data($0) } }
      func field(_ bytes: Data) { hash.update(data: number(UInt32(bytes.count))); hash.update(data: bytes) }
      field(Data(kind.domain.utf8))
      field(number(value.transport)); field(try copyBytes(value.host)); field(try copyBytes(value.scope)); field(number(value.port))
      field(try copyBytes(value.path)); field(try copyBytes(value.route))
      return "v1:" + hash.finalize().map { String(format: "%02x", $0) }.joined()
    }
    self.endpoint = endpoint; self.routeIdentity = routeIdentity; self.kind = kind
  }
}
public struct NativeTrustRevision: Sendable, Equatable, CustomStringConvertible, CustomDebugStringConvertible {
  fileprivate let digest: Data?
  public var description: String { "NativeTrustRevision(<redacted>)" }
  public var debugDescription: String { description }
  fileprivate init(_ data: Data?) { digest = data.map { Data(SHA256.hash(data: $0)) } }
}
public struct NativeSavedTrustEntry: Sendable, Identifiable, CustomStringConvertible, CustomDebugStringConvertible {
  public var id: String { scope.id }
  public let scope: NativeTrustScope
  // nil means forgotten: do not fall back to an older broad host exception.
  let identity: Data?
  public var fingerprint: String? { identity.map { NativeLegacyTrustCodec.fingerprint($0) } }
  public var isForgotten: Bool { identity == nil }
  public var description: String { "NativeSavedTrustEntry(<redacted>)" }
  public var debugDescription: String { description }
}
public struct NativeSavedTrustSnapshot: Sendable {
  public let revision: NativeTrustRevision
  public let entries: [NativeSavedTrustEntry]
}
public struct NativeSavedTrustInspection: Sendable, CustomStringConvertible, CustomDebugStringConvertible {
  public var description: String { "NativeSavedTrustInspection(<redacted>)" }
  public var debugDescription: String { description }
  public enum State: Sendable { case absent, forgotten, match, changed }
  public let state: State, revision: NativeTrustRevision
  public let expectedFingerprint: String?, receivedFingerprint: String
}
public struct NativeTrustCommit: Sendable {
  public let snapshot: NativeSavedTrustSnapshot
  // Replacement is observed on disk, but a post-rename sync failed. Never
  // translate this to cancellation or retry the write without a fresh read.
  public let durabilityUncertain: Bool
}
public struct NativeTrustFile: NativeAtomicFileBacking {
  public let kind: NativeTrustKind
  public init(kind: NativeTrustKind = .certificate) { self.kind = kind }
  public static func file(kind: NativeTrustKind = .certificate, environment: [String:String] = NativePathEnvironment.capture(),
                          home: URL = FileManager.default.homeDirectoryForCurrentUser) throws -> NativePrivateFile {
    let legacy = try NativeLegacyTrustFile.applicationStore(environment: environment,home: home)
    let directory = legacy.url.deletingLastPathComponent().appendingPathComponent("native-trust",isDirectory: true)
    return try NativePrivateFile(directory: directory,record: kind == .certificate ? .trustExceptions : .serverKeys,prepareParent: true,checkpoint: { _ in })
  }
  public func read() throws -> Data? { try Self.file(kind: kind).read() }
  public func replace(_ data: Data, expected: Data?) throws { try Self.file(kind: kind).replace(data,expected: expected) }
}
public actor NativeTrustStore {
  public static let capacity = 256
  public nonisolated let kind: NativeTrustKind
  private struct Entry: Codable {
    let scope: String, endpoint: String, route: String, decision: String
    let spki: Data?, hostKey: Data?
    var identity: Data? { spki ?? hostKey }
  }
  private struct Record: Codable { let schema: UInt32; let kind: String?; var entries: [Entry] }
  private let backing: any NativeAtomicFileBacking
  private let makeKey: @Sendable (Data) throws -> any NativeCertificateKeyMaterial
  private var closed = false
  public init(kind: NativeTrustKind = .certificate, backing: (any NativeAtomicFileBacking)? = nil,
              makeKey: @escaping @Sendable (Data) throws -> any NativeCertificateKeyMaterial = { try NativeCertificateKey(certificate: $0) }) {
    self.kind = kind; self.backing = backing ?? NativeTrustFile(kind: kind); self.makeKey = makeKey
  }
  private func admission() throws {
    guard !closed else { throw NativeStorageError.closed }
    guard !Task.isCancelled else { throw NativeStorageError.cancelled }
  }
  private func snapshot(_ record: Record, data: Data?) throws -> NativeSavedTrustSnapshot {
    guard record.entries.count <= Self.capacity else { throw NativeStorageError.resourceLimit }
    var seen = Set<String>(), entries: [NativeSavedTrustEntry] = []
    for entry in record.entries {
      let scope = try NativeTrustScope(endpoint: entry.endpoint,routeIdentity: entry.route,kind: kind)
      guard scope.id == entry.scope, seen.insert(scope.id).inserted else { throw NativeStorageError.corrupt }
      guard entry.decision == "accept" || entry.decision == "forget" else { throw NativeStorageError.unsupportedValue }
      if entry.decision == "accept" {
        guard let key = entry.identity, !key.isEmpty, key.count <= 65536 else { throw NativeStorageError.corrupt }
        if kind == .hostKey { do { _ = try NativeHostKey(key) } catch { throw NativeStorageError.corrupt } }
      } else if entry.identity != nil { throw NativeStorageError.corrupt }
      entries.append(.init(scope: scope,identity: entry.identity))
    }
    return .init(revision: NativeTrustRevision(data),entries: entries)
  }
  private func load() throws -> (Record,NativeSavedTrustSnapshot,Data?) {
    let data = try backing.read()
    guard let data else {
      let record = Record(schema: 1,kind: kind == .certificate ? nil : kind.rawValue,entries: [])
      return (record,try snapshot(record,data: nil),nil)
    }
    guard data.count <= NativePrivateFile.maximumBytes else { throw NativeStorageError.tooLarge }
    do {
      guard let root = try JSONSerialization.jsonObject(with: data) as? [String:Any],
            let schema = root["schema"] as? NSNumber, CFGetTypeID(schema) != CFBooleanGetTypeID(),
            schema.doubleValue >= 1, schema.doubleValue.rounded() == schema.doubleValue else { throw NativeStorageError.corrupt }
      guard schema.doubleValue == 1 else { throw NativeStorageError.futureSchema }
      let rootFields: Set<String> = kind == .certificate ? ["schema","entries"] : ["schema","kind","entries"]
      guard Set(root.keys) == rootFields else { throw NativeStorageError.unsupportedFields }
      if kind == .hostKey, root["kind"] as? String != kind.rawValue { throw NativeStorageError.unsupportedValue }
      guard let entries = root["entries"] as? [[String:Any]] else { throw NativeStorageError.corrupt }
      guard entries.count <= Self.capacity else { throw NativeStorageError.resourceLimit }
      for entry in entries {
        let required: Set<String> = ["scope","endpoint","route","decision"]
        guard required.isSubset(of: Set(entry.keys)), Set(entry.keys).isSubset(of: required.union([kind == .certificate ? "spki" : "hostKey"])) else { throw NativeStorageError.unsupportedFields }
        guard !entry.values.contains(where: { $0 is NSNull }) else { throw NativeStorageError.corrupt }
      }
      let record = try JSONDecoder().decode(Record.self,from: data)
      return (record,try snapshot(record,data: data),data)
    } catch let error as NativeStorageError { throw error }
    catch { throw NativeStorageError.corrupt }
  }
  public func read() throws -> NativeSavedTrustSnapshot { try admission(); return try load().1 }
  public func inspect(scope: NativeTrustScope, certificate: Data) throws -> NativeSavedTrustInspection {
    try admission()
    guard kind == .certificate, scope.kind == kind else { throw NativeStorageError.invalid }
    return try inspectIdentity(scope: scope,identity: makeKey(certificate).spki)
  }
  public func inspectHostKey(scope: NativeTrustScope, key: Data) throws -> NativeSavedTrustInspection {
    try admission()
    guard kind == .hostKey, scope.kind == kind else { throw NativeStorageError.invalid }
    return try inspectIdentity(scope: scope,identity: NativeHostKey(key).identity)
  }
  private func inspectIdentity(scope: NativeTrustScope,identity: Data) throws -> NativeSavedTrustInspection {
    let snapshot = try load().1, entry = snapshot.entries.first { $0.id == scope.id }
    return .init(state: entry == nil ? .absent : entry!.isForgotten ? .forgotten : entry!.identity == identity ? .match : .changed,
      revision: snapshot.revision,expectedFingerprint: entry?.fingerprint,receivedFingerprint: NativeLegacyTrustCodec.fingerprint(identity))
  }
  private func mutate(scope: NativeTrustScope, key: Data?, replacing: Bool, expected: NativeTrustRevision) throws -> NativeTrustCommit {
    try admission()
    guard scope.kind == kind else { throw NativeStorageError.invalid }
    var (record, original, bytes) = try load()
    guard original.revision == expected else { throw NativeStorageError.conflict }
    let index = record.entries.firstIndex { $0.scope == scope.id }
    let hasKey = index.map { record.entries[$0].identity != nil } ?? false
    if key != nil { guard replacing == hasKey else { throw NativeStorageError.conflict } }
    let entry = Entry(scope: scope.id,endpoint: scope.endpoint,route: scope.routeIdentity,decision: key == nil ? "forget" : "accept",spki: kind == .certificate ? key : nil,hostKey: kind == .hostKey ? key : nil)
    if let index { record.entries[index] = entry } else { record.entries.append(entry) }
    _ = try snapshot(record,data: nil)
    let encoder = JSONEncoder(); encoder.outputFormatting = [.sortedKeys]
    let encoded = try encoder.encode(record)
    guard encoded.count <= NativePrivateFile.maximumBytes else { throw NativeStorageError.tooLarge }
    let result = try snapshot(record,data: encoded)
    try admission()
    do { try backing.replace(encoded,expected: bytes) }
    catch {
      // A failed post-rename sync cannot be rolled back. Reconcile exact bytes;
      // no automatic retry, no overwrite of a third writer's later decision.
      if (try? backing.read()) == encoded { return .init(snapshot: result,durabilityUncertain: true) }
      throw error
    }
    return .init(snapshot: result,durabilityUncertain: false)
  }
  public func save(scope: NativeTrustScope, certificate: Data, status: UInt32, replacing: Bool,
                   expected: NativeTrustRevision) throws -> NativeTrustCommit {
    try admission()
    guard kind == .certificate, scope.kind == kind, try NativeCertificatePolicy(status: status).mayOverride else { throw NativeStorageError.invalid }
    let key = try makeKey(certificate).spki
    guard !key.isEmpty, key.count <= 65536 else { throw NativeStorageError.invalid }
    return try mutate(scope: scope,key: key,replacing: replacing,expected: expected)
  }
  public func saveHostKey(scope: NativeTrustScope, key: Data, replacing: Bool,
                          expected: NativeTrustRevision) throws -> NativeTrustCommit {
    try admission()
    guard kind == .hostKey, scope.kind == kind else { throw NativeStorageError.invalid }
    return try mutate(scope: scope,key: NativeHostKey(key).identity,replacing: replacing,expected: expected)
  }
  public func forget(scope: NativeTrustScope, expected: NativeTrustRevision) throws -> NativeTrustCommit {
    try mutate(scope: scope,key: nil,replacing: false,expected: expected)
  }
  public func close() { closed = true }
}

File: platform/windows/TidyVNC.Native/Trust/NativeTrustStore.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.Collections.Immutable;
using System.Security.Cryptography;
using System.Text.Json;
using TidyVNC.Native.Storage;

namespace TidyVNC.Native.Trust;

/// <summary>
/// The destination a trust decision applies to: the core's domain-separated
/// identity of the canonical endpoint and route (NativeTrustScope on macOS).
/// Labels are kept for the management window only; stored scopes are
/// rederived on load and never trusted as labels.
/// </summary>
public sealed class NativeTrustScope : IEquatable<NativeTrustScope>
{
    private NativeTrustScope(string id, string endpoint, string routeIdentity, NativeTrustKind kind)
    {
        Id = id; Endpoint = endpoint; RouteIdentity = routeIdentity; Kind = kind;
    }

    public string Id { get; }
    public string Endpoint { get; }
    public string RouteIdentity { get; }
    public NativeTrustKind Kind { get; }

    /// <summary>Throws <see cref="NativeIdentityFailure"/> for an empty, oversized or invalid endpoint.</summary>
    public static NativeTrustScope Create(string endpoint, string routeIdentity = "", NativeTrustKind kind = NativeTrustKind.Certificate)
    {
        if (endpoint.Length == 0) throw new NativeIdentityFailure(NativeIdentityFailure.Problem.InvalidEndpoint);
        return new(NativeIdentity.TrustScope(endpoint, routeIdentity, kind), endpoint, routeIdentity, kind);
    }

    public bool Equals(NativeTrustScope? other) => other is not null && string.Equals(Id, other.Id, StringComparison.Ordinal);
    public override bool Equals(object? obj) => Equals(obj as NativeTrustScope);
    public override int GetHashCode() => StringComparer.Ordinal.GetHashCode(Id);
    public override string ToString() => "NativeTrustScope(<redacted>)";
}

/// <summary>A saved decision; a null identity means forgotten (no fallback to a broader legacy exception).</summary>
public sealed class NativeSavedTrustEntry(NativeTrustScope scope, ImmutableArray<byte>? identity)
{
    public NativeTrustScope Scope { get; } = scope;
    public string Id => Scope.Id;
    public ImmutableArray<byte>? Identity { get; } = identity;
    public bool IsForgotten => Identity is null;
    public string? Fingerprint => Identity is { } key ? NativeTrustFingerprint.Of(key.AsSpan()) : null;
    public override string ToString() => "NativeSavedTrustEntry(<redacted>)";
}

public static class NativeTrustFingerprint
{
    /// <summary>SHA-256, upper-case hex pairs separated by colons (the dialog's and library's format).</summary>
    public static string Of(ReadOnlySpan<byte> bytes)
        => string.Join(':', SHA256.HashData(bytes).ToArray().Select(b => b.ToString("X2", System.Globalization.CultureInfo.InvariantCulture)));
}

public enum NativeSavedTrustState { Absent, Forgotten, Match, Changed }

public sealed record NativeSavedTrustInspection(NativeSavedTrustState State, Guid? Revision, string? SavedFingerprint, string ReceivedFingerprint)
{
    public override string ToString() => $"NativeSavedTrustInspection({State}, <redacted>)";
}

public sealed record NativeSavedTrustSnapshot(Guid? Revision, ImmutableArray<NativeSavedTrustEntry> Entries);

/// <summary>
/// trust\certificates.json (X.509 SPKI decisions) or trust\server-keys.json
/// (RSA-AES server keys): destination-scoped accept/forget decisions with the
/// macOS NativeTrustStore's kinds, capacity and matching rules (SERVICES.md
/// section 5), on the shared record store (schema 1, revision, lock, atomic
/// replace, owner-only DACL). No certificate ever enters a Windows store.
/// </summary>
public sealed class NativeTrustStore : NativeRecordStore<ImmutableArray<NativeSavedTrustEntry>>
{
    public const int Capacity = 256;
    public const int MaximumIdentityBytes = 65536;

    public NativeTrustStore(NativeTrustKind kind, string stateDirectory)
        : base(System.IO.Path.Combine(stateDirectory, "trust"), kind == NativeTrustKind.Certificate ? "certificates.json" : "server-keys.json")
    {
        Kind = kind;
    }

    public NativeTrustStore(NativeTrustKind kind) : this(kind, NativeStateRoot.Directory) { }

    public NativeTrustKind Kind { get; }
    private string KindName => Kind == NativeTrustKind.Certificate ? "x509-spki" : "rsa-aes";
    private string IdentityField => Kind == NativeTrustKind.Certificate ? "spki" : "hostKey";

    protected override int Schema => 1;
    protected override ImmutableArray<NativeSavedTrustEntry> Empty => [];

    protected override ImmutableArray<NativeSavedTrustEntry> Decode(JsonElement record, int schema)
    {
        RequireOnly(record, "kind", "entries");
        if (!record.TryGetProperty("kind", out var kind) || kind.ValueKind != JsonValueKind.String) throw Corrupt();
        if (kind.GetString() != KindName) throw new NativeStorageException(NativeStorageError.UnsupportedFields);
        if (!record.TryGetProperty("entries", out var list) || list.ValueKind != JsonValueKind.Array) throw Corrupt();
        if (list.GetArrayLength() > Capacity) throw new NativeStorageException(NativeStorageError.ResourceLimit);
        var entries = ImmutableArray.CreateBuilder<NativeSavedTrustEntry>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var item in list.EnumerateArray())
        {
            RequireOnly(item, "scope", "endpoint", "route", "decision", IdentityField);
            string Text(string name) => item.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString()! : throw Corrupt();
            NativeTrustScope scope;
            try { scope = NativeTrustScope.Create(Text("endpoint"), Text("route"), Kind); }
            catch (NativeIdentityFailure) { throw Corrupt(); }
            if (scope.Id != Text("scope") || !seen.Add(scope.Id)) throw Corrupt();
            var decision = Text("decision");
            ImmutableArray<byte>? identity = null;
            if (decision == "accept")
            {
                if (!item.TryGetProperty(IdentityField, out var encoded) || encoded.ValueKind != JsonValueKind.String) throw Corrupt();
                var bytes = encoded.GetBytesFromBase64();
                if (bytes.Length is 0 or > MaximumIdentityBytes || Convert.ToBase64String(bytes) != encoded.GetString()) throw Corrupt();
                if (Kind == NativeTrustKind.HostKey) ValidHostKey(bytes, NativeStorageError.Corrupt);
                identity = [.. bytes];
            }
            else if (decision == "forget")
            {
                if (item.TryGetProperty(IdentityField, out _)) throw Corrupt();
            }
            else throw new NativeStorageException(NativeStorageError.UnsupportedFields);
            entries.Add(new NativeSavedTrustEntry(scope, identity));
        }
        return entries.ToImmutable();
    }

    protected override void Encode(Utf8JsonWriter writer, ImmutableArray<NativeSavedTrustEntry> value)
    {
        if (value.Length > Capacity) throw new NativeStorageException(NativeStorageError.ResourceLimit);
        writer.WriteString("kind", KindName);
        writer.WriteStartArray("entries");
        foreach (var entry in value)
        {
            writer.WriteStartObject();
            writer.WriteString("scope", entry.Scope.Id);
            writer.WriteString("endpoint", entry.Scope.Endpoint);
            writer.WriteString("route", entry.Scope.RouteIdentity);
            writer.WriteString("decision", entry.IsForgotten ? "forget" : "accept");
            if (entry.Identity is { } identity) writer.WriteBase64String(IdentityField, identity.AsSpan());
            writer.WriteEndObject();
        }
        writer.WriteEndArray();
    }

    private static void ValidHostKey(byte[] key, NativeStorageError failure)
    {
        try { _ = new NativeHostKey(key); }
        catch (Exception error) when (error is NativeError or ArgumentException) { throw new NativeStorageException(failure); }
    }

    private void RequireKind(NativeTrustScope scope, NativeTrustKind kind)
    {
        if (Kind != kind || scope.Kind != Kind) throw new NativeStorageException(NativeStorageError.UnsupportedFields);
    }

    private static byte[] Spki(ReadOnlySpan<byte> certificate)
    {
        if (certificate.IsEmpty || certificate.Length > MaximumIdentityBytes) throw new NativeStorageException(NativeStorageError.Corrupt);
        try
        {
            using var key = new NativeCertificateKey(certificate);
            return key.Spki.Length is > 0 and <= MaximumIdentityBytes ? key.Spki : throw new NativeStorageException(NativeStorageError.Corrupt);
        }
        catch (NativeError) { throw new NativeStorageException(NativeStorageError.Corrupt); }
    }

    public async Task<NativeSavedTrustSnapshot> ReadSnapshotAsync(CancellationToken cancellation = default)
    {
        var snapshot = await ReadAsync(cancellation).ConfigureAwait(false);
        return new(snapshot.Revision, snapshot.Value);
    }

    /// <summary>Compares the presented certificate's SPKI with the saved decision for this destination.</summary>
    public async Task<NativeSavedTrustInspection> InspectAsync(NativeTrustScope scope, byte[] certificate, CancellationToken cancellation = default)
    {
        RequireKind(scope, NativeTrustKind.Certificate);
        return await Inspect(scope, Spki(certificate), cancellation).ConfigureAwait(false);
    }

    public async Task<NativeSavedTrustInspection> InspectHostKeyAsync(NativeTrustScope scope, byte[] key, CancellationToken cancellation = default)
    {
        RequireKind(scope, NativeTrustKind.HostKey);
        ValidHostKey(key, NativeStorageError.Corrupt);
        return await Inspect(scope, key, cancellation).ConfigureAwait(false);
    }

    private async Task<NativeSavedTrustInspection> Inspect(NativeTrustScope scope, byte[] identity, CancellationToken cancellation)
    {
        var snapshot = await ReadAsync(cancellation).ConfigureAwait(false);
        var entry = snapshot.Value.FirstOrDefault(e => e.Scope.Equals(scope));
        var state = entry is null ? NativeSavedTrustState.Absent
            : entry.Identity is not { } saved ? NativeSavedTrustState.Forgotten
            : saved.AsSpan().SequenceEqual(identity) ? NativeSavedTrustState.Match : NativeSavedTrustState.Changed;
        return new(state, snapshot.Revision, entry?.Fingerprint, NativeTrustFingerprint.Of(identity));
    }

    /// <summary>
    /// Saves an exception for a certificate whose verification status may be
    /// overridden. replacing states that the caller showed a changed key;
    /// it must agree with the record or the result is a Conflict.
    /// </summary>
    public Task<NativeSavedTrustSnapshot> SaveAsync(NativeTrustScope scope, byte[] certificate, uint status, bool replacing, Guid? expected,
                                                    CancellationToken cancellation = default)
    {
        RequireKind(scope, NativeTrustKind.Certificate);
        if (!NativeCertificatePolicy.For(status).MayOverride) throw new NativeStorageException(NativeStorageError.Denied);
        return Mutate(scope, [.. Spki(certificate)], replacing, expected, cancellation);
    }

    public Task<NativeSavedTrustSnapshot> SaveHostKeyAsync(NativeTrustScope scope, byte[] key, bool replacing, Guid? expected,
                                                           CancellationToken cancellation = default)
    {
        RequireKind(scope, NativeTrustKind.HostKey);
        ValidHostKey(key, NativeStorageError.UnsupportedFields);
        return Mutate(scope, [.. key], replacing, expected, cancellation);
    }

    /// <summary>Records a forget decision, which also suppresses the legacy fallback for this destination.</summary>
    public Task<NativeSavedTrustSnapshot> ForgetAsync(NativeTrustScope scope, Guid? expected, CancellationToken cancellation = default)
    {
        RequireKind(scope, Kind);
        return Mutate(scope, null, false, expected, cancellation);
    }

    private async Task<NativeSavedTrustSnapshot> Mutate(NativeTrustScope scope, ImmutableArray<byte>? identity, bool replacing, Guid? expected,
                                                        CancellationToken cancellation)
    {
        var current = await ReadAsync(cancellation).ConfigureAwait(false);
        if (current.Revision != expected) throw new NativeStorageException(NativeStorageError.Conflict);
        var index = -1;
        for (var i = 0; i < current.Value.Length && index < 0; i++)
            if (current.Value[i].Scope.Equals(scope)) index = i;
        var hasKey = index >= 0 && !current.Value[index].IsForgotten;
        if (identity is not null && replacing != hasKey) throw new NativeStorageException(NativeStorageError.Conflict);
        var entry = new NativeSavedTrustEntry(scope, identity);
        var entries = index >= 0 ? current.Value.SetItem(index, entry) : current.Value.Add(entry);
        if (entries.Length > Capacity) throw new NativeStorageException(NativeStorageError.ResourceLimit);
        var committed = await CommitAsync(entries, expected, cancellation).ConfigureAwait(false);
        return new(committed.Revision, committed.Value);
    }
}
